src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { PrincipalType } from '../../../../core/utils/constants';
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Partial<Record<string, Principals>>;
+
+const addPermissionToPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deletePermissionFromPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions;
+  }
+
+  // parse the permissions object to check whether the specific user or group has the specific permission or not
+  public hasPermission(
+    permissionType: string,
+    permissions: Permissions,
+    user?: string,
+    group?: string
```

Comment:
Will we make these two params as a single parameters like principles? Or we need to modify all the methods once there is a new principle type comes up. And I think all the principles here should be a string[] because a user may have multiple backend roles.